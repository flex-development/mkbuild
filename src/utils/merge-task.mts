/**
 * @file Utilities - mergeTask
 * @module mkbuild/utils/mergeTask
 */

import type { Task } from '@flex-development/mkbuild'
import {
  isObjectCurly,
  ksort,
  properties,
  sift
} from '@flex-development/tutils'

export default mergeTask

/**
 * Merge one or more build tasks into a single {@linkcode Task}.
 *
 * Task source objects are applied from left to right. Subsequent sources
 * overwrite property assignments of previous sources.
 *
 * @see {@linkcode Task}
 *
 * @category
 *  utils
 *
 * @this {void}
 *
 * @param {Task | null | undefined} target
 *  Target build task
 * @param {ReadonlyArray<Task | null | undefined>} tasks
 *  Source build task(s)
 * @return {Task}
 *  Merged build task
 */
function mergeTask(
  this: void,
  target: Task | null | undefined,
  ...tasks: readonly (Task | null | undefined)[]
): Task {
  return ksort(merge(target ?? {}, ...sift(tasks)))
}

/**
 * @this {void}
 *
 * @param {Record<PropertyKey, any>} target
 *  Target object
 * @param {Record<PropertyKey, any>[]} sources
 *  Source object(s)
 * @return {Record<PropertyKey, any>}
 *  Merged object
 */
function merge(
  this: void,
  target: Record<PropertyKey, any>,
  ...sources: Record<PropertyKey, any>[]
): Record<PropertyKey, any> {
  return sources.reduce<Record<PropertyKey, any>>((acc, source) => {
    return properties(source).reduce((target, property) => {
      /**
       * Source object value.
       *
       * @var {any} value
       */
      let value: any = source[property]

      if (Object.prototype.hasOwnProperty.call(target, property)) {
        /**
         * Target object value.
         *
         * @const {any} targetValue
         */
        const targetValue: any = target[property]

        if (Array.isArray(targetValue) && Array.isArray(value)) {
          value = [...new Set([...targetValue, ...value])]
        } else if (targetValue instanceof Set && value instanceof Set) {
          value = new Set([...targetValue, ...value])
        } else if (isObjectCurly(targetValue) && isObjectCurly(value)) {
          value = merge(targetValue, value)
        }
      }

      return target[property] = value, target
    }, acc)
  }, target)
}
