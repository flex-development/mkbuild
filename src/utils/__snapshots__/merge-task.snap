// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:utils/mergeTask > should return merged build task (0) 1`] = `
{
  "input": [
    "src/*.mts",
  ],
  "root": "__fixtures__/pkg/reverse",
  "sourcemap": true,
}
`;

exports[`unit:utils/mergeTask > should return merged build task (1) 1`] = `
{
  "esbuild": {
    "logOverride": {
      "ambiguous-reexport": "warning",
      "assert-to-with": "warning",
      "assert-type-json": "warning",
      "assign-to-constant": "warning",
      "assign-to-define": "warning",
      "assign-to-import": "warning",
      "call-import-namespace": "warning",
      "class-name-will-throw": "warning",
      "commonjs-variable-in-esm": "warning",
      "css-syntax-error": "warning",
      "delete-super-property": "warning",
      "different-path-case": "warning",
      "direct-eval": "warning",
      "duplicate-case": "warning",
      "duplicate-class-member": "warning",
      "duplicate-object-key": "warning",
      "empty-glob": "warning",
      "empty-import-meta": "warning",
      "equals-nan": "warning",
      "equals-negative-zero": "warning",
      "equals-new-object": "warning",
      "html-comment-in-js": "warning",
      "ignored-bare-import": "warning",
      "ignored-dynamic-import": "warning",
      "import-is-undefined": "warning",
      "impossible-typeof": "warning",
      "indirect-require": "warning",
      "invalid-@charset": "warning",
      "invalid-@import": "warning",
      "invalid-@layer": "warning",
      "invalid-calc": "warning",
      "invalid-source-mappings": "warning",
      "js-comment-in-css": "warning",
      "missing-source-map": "warning",
      "package.json": "warning",
      "private-name-will-throw": "warning",
      "require-resolve-not-external": "warning",
      "sections-in-source-map": "warning",
      "semicolon-after-return": "warning",
      "suspicious-boolean-not": "warning",
      "suspicious-define": "warning",
      "suspicious-logical-operator": "warning",
      "suspicious-nullish-coalescing": "warning",
      "this-is-undefined-in-esm": "warning",
      "tsconfig.json": "warning",
      "undefined-composes-from": "warning",
      "unsupported-@charset": "warning",
      "unsupported-@namespace": "warning",
      "unsupported-css-nesting": "warning",
      "unsupported-css-property": "warning",
      "unsupported-dynamic-import": "silent",
      "unsupported-jsx-comment": "warning",
      "unsupported-regexp": "warning",
      "unsupported-require-call": "warning",
      "unsupported-source-map-comment": "warning",
    },
    "target": [
      "es2024",
      "node18",
    ],
  },
  "input": [
    "src/*.mts",
    "src/interfaces/*.mts",
    "src/types/*.mts",
    "src/utils/*.mts",
  ],
  "resolve": {
    "conditions": Set {
      "import",
      "node",
      "mkbuild",
      "production",
    },
  },
  "tsconfig": "tsconfig.build.json",
}
`;
